File: src/GitOpen.Ui/Components/BranchTree.cs
using GitOpen.Domain.Refs;

namespace GitOpen.Ui.Components;

public sealed class BranchTreeNode
{
    public required string Name { get; init; }
    public required string FullPath { get; init; }
    public Branch? Branch { get; init; }
    public List<BranchTreeNode> Children { get; } = new();
    public bool IsLeaf => Branch is not null && Children.Count == 0;
}

public static class BranchTree
{
    public static List<BranchTreeNode> Build(IEnumerable<Branch> branches)
    {
        var roots = new List<BranchTreeNode>();
        var lookup = new Dictionary<string, BranchTreeNode>(StringComparer.Ordinal);

        foreach (var branch in branches)
        {
            var parts = branch.Name.Split('/');
            BranchTreeNode? parent = null;
            var currentPath = string.Empty;

            for (var i = 0; i < parts.Length; i++)
            {
                currentPath = i == 0 ? parts[0] : $"{currentPath}/{parts[i]}";
                var isLast = i == parts.Length - 1;

                if (!lookup.TryGetValue(currentPath, out var node))
                {
                    node = new BranchTreeNode
                    {
                        Name = parts[i],
                        FullPath = currentPath,
                        Branch = isLast ? branch : null
                    };
                    lookup[currentPath] = node;
                    if (parent is null) roots.Add(node);
                    else parent.Children.Add(node);
                }
                parent = node;
            }
        }

        SortRecursive(roots);
        return roots;
    }

    private static void SortRecursive(List<BranchTreeNode> nodes)
    {
        nodes.Sort((a, b) =>
        {
            var aIsFolder = a.Children.Count > 0;
            var bIsFolder = b.Children.Count > 0;
            if (aIsFolder != bIsFolder) return aIsFolder ? -1 : 1;
            return string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase);
        });
        foreach (var n in nodes) SortRecursive(n.Children);
    }

    public static IEnumerable<string> AllFolderPaths(IEnumerable<BranchTreeNode> nodes)
    {
        foreach (var n in nodes)
        {
            if (n.Children.Count > 0)
            {
                yield return n.FullPath;
                foreach (var sub in AllFolderPaths(n.Children)) yield return sub;
            }
        }
    }
}

File: lib/ui/sidebar/branch_tree.dart
import '../../domain/refs/branch.dart';

final class BranchTreeNode {
  final String name;
  final String fullPath;
  final Branch? branch;
  final List<BranchTreeNode> children;

  BranchTreeNode({required this.name, required this.fullPath, this.branch})
      : children = [];

  bool get isLeaf => branch != null && children.isEmpty;
}

class BranchTree {
  static List<BranchTreeNode> build(Iterable<Branch> branches) {
    final roots = <BranchTreeNode>[];
    final lookup = <String, BranchTreeNode>{};

    for (final b in branches) {
      final parts = b.name.split('/');
      BranchTreeNode? parent;
      var currentPath = '';
      for (var i = 0; i < parts.length; i++) {
        currentPath = i == 0 ? parts[0] : '$currentPath/${parts[i]}';
        final isLast = i == parts.length - 1;
        var node = lookup[currentPath];
        if (node == null) {
          node = BranchTreeNode(
            name: parts[i],
            fullPath: currentPath,
            branch: isLast ? b : null,
          );
          lookup[currentPath] = node;
          if (parent == null) {
            roots.add(node);
          } else {
            parent.children.add(node);
          }
        }
        parent = node;
      }
    }
    _sortRecursive(roots);
    return roots;
  }

  static void _sortRecursive(List<BranchTreeNode> nodes) {
    nodes.sort((a, b) {
      final aFolder = a.children.isNotEmpty;
      final bFolder = b.children.isNotEmpty;
      if (aFolder != bFolder) return aFolder ? -1 : 1;
      return a.name.toLowerCase().compareTo(b.name.toLowerCase());
    });
    for (final n in nodes) {
      _sortRecursive(n.children);
    }
  }

  static Iterable<String> allFolderPaths(Iterable<BranchTreeNode> nodes) sync* {
    for (final n in nodes) {
      if (n.children.isNotEmpty) {
        yield n.fullPath;
        yield* allFolderPaths(n.children);
      }
    }
  }
}
